device/transport/amqp/src/amqp_twin_client.ts

```diff
@@ -136,164 +169,96 @@ export class AmqpTwinClient extends EventEmitter {
-            this._appendEventQueue();
-          },
-          handleRemoveListener: () => {
-            this._appendEventQueue();
-          },
-          sendTwinRequest: () => {
-            this._appendEventQueue();
-          }
+          '*': () => this._fsm.deferUntilTransition()
         }
       }
     });
+  }
 
-    this._fsm.on('transition', (transition) => {
-      debug(transition.fromState + ' -> ' + transition.toState + ' (action:' + transition.action + ')');
-    });
-
-    this.on('newListener', this._handleNewListener.bind(this));
-    this.on('removeListener', this._handleRemoveListener.bind(this));
-    this._boundMessageHandler = this._onAmqpMessage.bind(this); // need to save this so that calls to add & remove listeners can be matched by the EventEmitter.
+  getTwin(callback: (err: Error, twin?: TwinProperties) => void): void {
+    this._fsm.handle('getTwin', callback);
+  }
 
+  updateTwinReportedProperties(patch: any, callback: (err?: Error) => void): void {
+    this._fsm.handle('updateTwinReportedProperties', patch, callback);
   }
 
-  /**
-   * @method          module:azure-iot-device-amqp.Amqp#sendTwinRequest
-   * @description     Send a device-twin specific message to the IoT Hub instance
-   *
-   * @param {String}        method    name of the method to invoke ('PUSH', 'PATCH', etc)
-   * @param {String}        resource  name of the resource to act on (e.g. '/properties/reported/') with beginning and ending slashes
-   * @param {Object}        properties  object containing name value pairs for request properties (e.g. { 'rid' : 10, 'index' : 17 })
-   * @param {String}        body  body of request
-   * @param {Function}      done  the callback to be invoked when this function completes.
-   *
-   * @throws {ReferenceError}   One of the required parameters is falsy
-   * @throws {ArgumentError}  One of the parameters is an incorrect type
-   */
-  sendTwinRequest(method: string, resource: string, properties: { [key: string]: string }, body: any, done?: (err?: Error, result?: any) => void): void {
-    this._fsm.handle('sendTwinRequest', method, resource, properties, body, done);
+  enableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void {
+    this._fsm.handle('enableTwinDesiredPropertiesUpdates', callback);
   }
 
-  attach(callback: (err: Error) => void): void {
-    this._fsm.handle('attach', callback);
+  disableTwinDesiredPropertiesUpdates(callback: (err?: Error) => void): void {
+    this._fsm.handle('disableTwinDesiredPropertiesUpdates', callback);
   }
 
-  detach(callback: (err: Error) => void): void {
+  /**
+   * Necessary for the client to be able to properly detach twin links
+   * attach() isn't necessary because it's done by the FSM automatically when one of the APIs is called.
+   */
+  detach(callback: (err?: Error) => void): void {
     this._fsm.handle('detach', callback);
   }
 
+
   private _generateTwinLinkProperties( correlationId: string): any {
```

Comment:
>_generateTwinLinkProperties [](start = 10, length = 27)

Do my eyes deceive me, or is there whitespace above this line?

Reply:
damnit!

---
In reply to: [171688487](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171688487) [](ancestors = 171688487)